flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/BaumWelch.scala

```diff
@@ -0,0 +1,346 @@
+            }
+            collector.collect(("t", state1, state2, prob))
+          }
+        }
+
+        //update emission probabilities
+        for (state <- 0 until numberOfStates) {
+          for (obVocab <- 0 until sizeOfObservationVocab) {
+            var sumProb: Double = 0
+            var sumGamma: Double = 0
+            for (o <- 0 until in.length) {
+              probSum = sumProbOfObservation(f, b, o)
+              if (probSum != 0) {
+                var g = gamma(state, f, b, o, probSum)
+                if (obVocab == in(o)) {
+                  sumProb += g
+                }
+                sumGamma += g
+              }
+            }
+
+            var prob: Double = 0
+            if (sumGamma != 0) {
+              prob = sumProb / sumGamma
+            }
+            collector.collect(("e", state, obVocab, prob))
+          }
+        }
+      }
+    }).withBroadcastSet(previousHmmModelParams, "HMMParams")
+  }
+
+
+  /**
+   * combines the results of the distributed training
+   *
+   * @param results results of the distributed trainings
+   * @param numberOfStates number of states of the Hidden Markov Model
+   * @param sizeOfObservationVocab size of the obervation Vocabulary of the Hidden Markov Model
+   * @return new model parameters of the Hidden Markov Model
+   */
+  def combineResults(results: DataSet[(String, Int, Int, Double)], numberOfStates: Int,
+                     sizeOfObservationVocab: Int):
+  DataSet[(DenseVector, DenseMatrix, DenseMatrix)] = {
+    results.groupBy(0, 1, 2).sum(3).map(t => (t._1, t._2, List((t._3, t._4)), t._4))
+      .groupBy(0, 1).reduce((tuple1, tuple2) =>
+      (tuple1._1, tuple1._2, tuple1._3 ::: tuple2._3, tuple1._4 + tuple2._4))
+      .reduceGroup(iterator => {
+        var initialProbabilities: DenseVector = null
+        var transitionProbabilities: DenseMatrix = null
+        var emissionProbabilities: DenseMatrix = null
+        var count: Double = 0
+        iterator.foreach(p => {
+          if (p._1.equals("i")) {
+            if (initialProbabilities == null) {
+              initialProbabilities = new DenseVector(new Array[Double](numberOfStates))
+            }
+            initialProbabilities(p._2.toInt) = p._4
+            count += p._4
+          } else if (p._1.equals("t")) {
+            if (transitionProbabilities == null) {
+              transitionProbabilities = new DenseMatrix(numberOfStates, numberOfStates,
+                new Array[Double](numberOfStates * numberOfStates))
+            }
+            p._3.foreach(t => transitionProbabilities(p._2.toInt, t._1.toInt) = t._2 / p._4)
+          } else if (p._1.equals("e")) {
+            if (emissionProbabilities == null) {
+              emissionProbabilities = new DenseMatrix(numberOfStates, sizeOfObservationVocab,
+                new Array[Double](numberOfStates * sizeOfObservationVocab))
+            }
+            p._3.foreach(t => emissionProbabilities(p._2.toInt, t._1.toInt) = t._2 / p._4)
```

Comment:
You have to define all possible states and the vocabulary of all possible obervations before you train your HMM.
In future we can extend the HMM that it is not necessary to define all possible observations. Instead, you have to define a probability density function.